File: docs/src/test/java/com/example/javadsl/CustomBackendExample.java
package com.example.javadsl;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.checkpoint.CheckpointBackend;
import akka.stream.checkpoint.CheckpointRepository;
import akka.stream.checkpoint.javadsl.Checkpoint;
import akka.stream.javadsl.Source;

public class CustomBackendExample {

    public static void main(String[] args) {
        final ActorSystem system        = ActorSystem.create("DropwizardExample");
        final Materializer materializer = ActorMaterializer.create(system);

        // #custom
        final CheckpointBackend backend = new CheckpointBackend() {
            @Override
            public CheckpointRepository createRepository(String name) {
                return new CheckpointRepository() {
                    @Override
                    public void markPush(long latencyNanos, long backpressureRatio) {
                        System.out.println(String.format("PUSH - %s: latency:%d, backpressure ratio: %d",
                                name, latencyNanos, backpressureRatio));
                    }

                    @Override
                    public void markPull(long latencyNanos) {
                        System.out.println(String.format("PULL - %s: latency:%d",
                                name, latencyNanos));
                    }
                };
            }
        };

        Source.range(1, 100)
            .via(Checkpoint.create("produced", backend))
            .filter(x -> x % 2 == 0)
            .via(Checkpoint.create("filtered", backend))
            .runForeach(System.out::println, materializer);
        // #custom
    }
}

File: docs/src/test/scala/com/example/scaladsl/CustomBackendExample.scala
package com.example.scaladsl

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.checkpoint.{CheckpointBackend, CheckpointRepository}
import akka.stream.checkpoint.scaladsl.Checkpoint
import akka.stream.scaladsl.Source

object CustomBackendExample extends App {
  implicit val system       = ActorSystem.create("CustomBackendScalaExample")
  implicit val materializer = ActorMaterializer.create(system)
  import system.dispatcher

  // #custom
  implicit val loggingBackend: CheckpointBackend = new CheckpointBackend {
    override def createRepository(name: String): CheckpointRepository = new CheckpointRepository {

      override def markPush(latencyNanos: Long, backpressureRatio: Long): Unit =
        println(s"PUSH - $name: latency:$latencyNanos, backpressure ratio:$backpressureRatio")

      override def markPull(latencyNanos: Long): Unit =
        println(s"PULL - $name: latency:$latencyNanos")
    }
  }

  Source(1 to 100)
    .via(Checkpoint("produced"))
    .filter((x: Int) ⇒ x % 2 == 0)
    .via(Checkpoint("filtered"))
    .runForeach(println)
  // #custom
    .onComplete { _ ⇒  system.terminate() }
}
